### PR TITLE
Finding sets

### LINKED ISSUE #5: Users can find sets.

### LINKED ISSUE #6: New cards are added when a set is accepted.

### PR DESCRIPTION
**Pull request by connorlay**
resolves #6  
resolves #5 

Users can submit sets to a game. If the set is accepted, the board is updated and the user is awarded a point. If the set is rejected, the user loses a point.
pr4 1gif
pr4 2gif

### CODE REVIEW

File: app/controllers/api/v1/games_controller.rb [hunk 1]
```diff
@@ -1,31 +1,35 @@
-module Api::V1
-  class GamesController < ApplicationController
+class Api::V1::GamesController < Api::V1::ApplicationController
 
-    before_action :conflict_error, if: -> { lobby.game.present? },
-                                   only: [ :create ]
-    before_action :check_user_membership, only: [ :show ]
+  before_action :conflict_error, if: -> { lobby.game.present? },
```
Comment (Fosome) [inline app/controllers/api/v1/games_controller.rb]: This is the same as ln 6?  Why are you doing this twice?

Comment (connorlay) [inline app/controllers/api/v1/games_controller.rb]: Completely forgot to remove that line after rebasing.

File: app/controllers/api/v1/games_controller.rb [hunk 2]
```diff
@@ -1,31 +1,35 @@
...
-      @game ||= Game.find(params[:id])
-    end
+  def game
+    @game ||= Game.find(params[:id])
+  end
 
-    def lobby
-      @lobby ||= Lobby.find(params[:lobby_id])
-    end
+  def check_user_membership
```
Comment (Fosome) [inline app/controllers/api/v1/games_controller.rb]: Can you share this across controllers?

File: app/controllers/api/v1/lobbies_controller.rb [hunk 3]
```diff
@@ -1,27 +1,24 @@
...
-      new_lobby = Lobby.create
-      new_lobby.add_user(current_user)
-      render json: new_lobby, serializer: LobbySerializer
-    end
-
-    def show
-      render json: lobby, serializer: LobbySerializer
-    end
+  def create
+    new_lobby = Lobbies::Creator.new.create_new_lobby(current_user)
```
Comment (Fosome) [inline app/controllers/api/v1/lobbies_controller.rb]: `Lobbies::Creator.by_user(current_user)` or similar.  This API is a little strange.

File: app/controllers/api/v1/lobbies_controller.rb [hunk 4]
```diff
@@ -1,27 +1,24 @@
...
 
-    def lobby
-      @lobby ||= Lobby.find(params[:id])
-    end
+  private
 
-    def check_user_membership
-      authentication_error unless lobby.has_user?(current_user)
-    end
+  def check_user_membership
```
Comment (Fosome) [inline app/controllers/api/v1/lobbies_controller.rb]: Dup

**Comment by Fosome** [comment]
The use of "service objects" here is questionable.  They should provide a simple interface for complex or multi-step actions.  This feels more like shifting around functions.  It's great thinking but needs more refinement.

File: app/controllers/api/v1/lobbies_controller.rb (hunk 3 above)
Comment (connorlay) [inline app/controllers/api/v1/lobbies_controller.rb]: I still want `create` in the method name. Hoes does `Lobbies::Creator.new.create_with_user(current_user)` sound?

Comment (Fosome) [inline app/controllers/api/v1/lobbies_controller.rb]: The name feels like more than enough context.  That aside, I have a larger issue with `Thinger.new.create(param)`.  Should either be a class method that accepts parameters or an object that manages state.  This is weirdly in between.

**Comment by connorlay** [comment]
@Fosome The service object design pattern is still new to me. Here's my main question: is the issue that the app itself is not that complex, so introducing new objects to encapsulate functionality is premature optimization, or more that the very idea of putting most application logic into POROs goes against the "Rails way" of models, controllers, helpers, and views?

Here's where I'm at so far:

- Models are for interacting directly with the database and should not know or care about application specific logic (like how to populate `deck` and `board`). Models are useful for validating data and creating associations.
- Controllers are the doorkeepers for the user to interact with the app. They handle accepting input and rendering views. Controllers should not implement application logic themselves. In other words, you should be able to interact with your app entirely through the rails console.
- Views are for presenting data. How to arrange for that data to be displayed is the job of the controller (although I've read about presenter objects as well). For this app, serializers are my views.
- Helpers are supposed to contain application logic that does not belong in the above three. The main issue I have with this is a) it's difficult to locate specific methods that magically appear elsewhere, and b) it breaks the single-responsibility-principle (_"My responsibility is to contain methods that don't belong anywhere else."_ is not focused enough).

I'd love to hear your thoughts on this!

**Comment by Fosome** [comment]
Everything you've said here is great.  Patterns tho, like service objects, are just patterns.  I think there is a difference here between seeing the pattern *emerge* and forcing its application.  

This is tough in Rails applications because Active Record (also a pattern) don't have a single responsibility and can quickly become god objects.  So in some sense,

```ruby
  # lobby_controller.rb

  def create
    @lobby = Lobby.new(lobby_params)
    if @lobby.save
      #...
    end
  end
```

`Lobby` is a service object.  All AR classes are service objects.  They can play this role well because they already have the biz logic (validations, etc).  They also expose the state of the operation on the instance it returns -- success, failure, pending.

So in the very least, it's probably worth creating PORO service objects that behave similarly.  Imagine one of your service objects *failing* its transaction.  What would that look like?

**Comment by connorlay** [comment]
I'd imagine something like `#success?` That way you can determine if the actions performed by the service object performed as expected. [This blog post](https://netguru.co/blog/service-objects-in-rails-will-help) suggests standardizing all service objects to have this kind of interface:
~~~ruby
service = MyServiceObject.new(params)
result = service.call
if result.success?
  # do stuff with result.data
else
  # handle errors here
end
~~~

The result of each service is wrapped in an object like so:
~~~ruby
class Success
  attr_reader :data

  def initialize(data)
    @data = data
  end

  def success?
    true
  end
end

class Error
  attr_reader :error

  def initialize(error)
    @error = error
  end

  def success?
    false
  end
end
~~~

A rewrite of `Lobbies::Creator` could look like this:
~~~ruby
result = Lobbies::Creator.new(user: current_user).call
if result.success?
  render json: result.data, serializer: LobbySerializer
else
  conflict_error
end
~~~
  
I agree that all of these POROs should share a common interface. I'm not sure if `.initialize(params)`, `#call`, and `#success` are clear enough?

**Comment by efatsi** [comment]
:+1:

File: app/models/game.rb [hunk 5]
```diff
@@ -4,32 +4,20 @@ class Game < ActiveRecord::Base
...
+  def remove_from_board(ids)
+    ids.each { |id| board.delete(id) }
+    save!
   end
 
-  def setup_game
-    self.deck  = dealer.new_deck.map(&:id)
-    self.board = read_attribute(:deck).pop(12)
+  def setup!
+    self.deck.shuffle!
```
Comment (Fosome) [inline app/models/game.rb]: don't need `self` here

File: lib/cards_table.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
+class CardsTable
+
+  attr_reader :cards, :ids
+
+  def initialize(args)
```
Comment (Fosome) [inline lib/cards_table.rb]: check out https://robots.thoughtbot.com/ruby-2-keyword-arguments

**Comment by Fosome** [comment]
:+1:

**Pull request closed without merging**